src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt

```diff
@@ -1119,24 +1101,28 @@ class Peer(
-                        // payment rather that having the user wonder where their money went.
-                        val channelInitializing = _channels.isNotEmpty()
-                                && !_channels.values.any { it is Normal } // we don't have a channel that can be spliced
-                                && _channels.values.any { it is WaitForFundingSigned || it is WaitForFundingConfirmed || it is WaitForChannelReady } // but we will have one soon
-                        if (channelInitializing) {
-                            val rejected = LiquidityEvents.Rejected(msg.amountMsat, msg.payToOpenFeeSatoshis.toMilliSatoshi(), LiquidityEvents.Source.OffChainPayment, LiquidityEvents.Rejected.Reason.ChannelInitializing)
-                            logger.info { "rejecting pay-to-open: reason=${rejected.reason}" }
-                            nodeParams._nodeEvents.emit(rejected)
-                            val action = IncomingPaymentHandler.actionForPayToOpenFailure(nodeParams.nodePrivateKey, TemporaryNodeFailure, msg)
-                            input.send(action)
-                        } else {
-                            processIncomingPayment(Either.Left(msg))
+                    is WillAddHtlc -> when {
+                        nodeParams.features.hasFeature(Feature.OnTheFlyFunding) -> when {
+                            nodeParams.liquidityPolicy.value is LiquidityPolicy.Disable -> {
+                                logger.warning { "cannot accept on-the-fly funding: policy set to disabled" }
+                                val failure = OutgoingPaymentPacket.buildWillAddHtlcFailure(nodeParams.nodePrivateKey, msg, TemporaryNodeFailure)
+                                input.send(SendOnTheFlyFundingMessage(failure))
+                                nodeParams._nodeEvents.emit(LiquidityEvents.Rejected(msg.amount, 0.msat, LiquidityEvents.Source.OffChainPayment, LiquidityEvents.Rejected.Reason.PolicySetToDisabled))
+                            }
+                            channelFundingIsInProgress() -> {
+                                // Once the channel funding is complete, we may have enough inbound liquidity to receive the payment without
+                                // an on-chain operation, which is more efficient. We thus reject that payment and wait for the sender to retry.
+                                logger.warning { "cannot accept on-the-fly funding: another funding attempt is already in-progress" }
+                                val failure = OutgoingPaymentPacket.buildWillAddHtlcFailure(nodeParams.nodePrivateKey, msg, TemporaryNodeFailure)
+                                input.send(SendOnTheFlyFundingMessage(failure))
+                                nodeParams._nodeEvents.emit(LiquidityEvents.Rejected(msg.amount, 0.msat, LiquidityEvents.Source.OffChainPayment, LiquidityEvents.Rejected.Reason.ChannelFundingInProgress))
+                            }
```

Comment:
I think in that case we could keep the `will_add_htlc` messages around, and replay them once the current splice is finished. This would create more risks around timeouts though, so we'll need to carefully think about this.